File: May-LeetCoding-Challenge/29-Course-Schedule/Course-Schedule.cpp
class Solution {
public:
   
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> deg(numCourses, 0);
        vector<vector<int>> G(numCourses, vector<int>());
        for (int i = 0; i < prerequisites.size(); i ++){
            deg[prerequisites[i][0]]++;
            G[prerequisites[i][1]].push_back(prerequisites[i][0]);
        }
        
        queue<int> q;
        for (int i = 0; i < numCourses; i++)  if (!deg[i]) q.push(i);
        int sum = 0;
        while(!q.empty()){
            int i = q.front(); 
            q.pop();
            sum ++;
            for (auto j: G[i]){
                deg[j]--;
                if (!deg[j]) q.push(j);
            }
        }
        
        return sum == numCourses;
        
    }
};

File: May-LeetCoding-Challenge/29-Course-Schedule/Course-Schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        List<Integer>[] adj = new ArrayList[numCourses];
        
        for (int i = 0; i < numCourses; i++) {
            adj[i] = new ArrayList();
        }
        
        for (int[] pre: prerequisites) {
            adj[pre[0]].add(pre[1]);
        }
        
        int[] visited = new int[numCourses];
        for (int i = 0; i < numCourses; i++) {
            if (visited[i] == 0 && !dfs(adj, visited, i)) {
                return false;
            }
        }
        
        return true;
    }
    
    private boolean dfs (List<Integer>[] adj, int[] visited, int v) {
        /**
         * 0 --> unvisited
         * 1 --> being visited
         * 2 --> completly visited
         */
        if (visited[v] == 1) {
            return false;
        }
        
        visited[v] = 1;
        
        for (int ad: adj[v]) {
            if (!dfs(adj, visited, ad)) return false;
        }
        
        visited[v] = 2;
        
        return true;
    }    
}